arch/x86_64/core/x86_64.c

```diff
@@ -0,0 +1,209 @@
+	cpu_init[cpu_num].fn = fn;
+}
+
+#ifdef CONFIG_IRQ_OFFLOAD
+static irq_offload_routine_t offload_fn;
+static void *offload_arg;
+
+static void irq_offload_handler(void *arg, int err)
+{
+	ARG_UNUSED(arg);
+	ARG_UNUSED(err);
+	offload_fn(offload_arg);
+}
+
+void irq_offload(irq_offload_routine_t fn, void *arg)
+{
+	offload_fn = fn;
+	offload_arg = arg;
+	__asm__ volatile("int %0" : : "i"(CONFIG_IRQ_OFFLOAD_VECTOR));
+}
+#endif
+
+/* Default.  Can be overridden at link time by a timer driver */
+void __weak x86_apic_timer_isr(void *arg, int code)
+{
+	ARG_UNUSED(arg);
+	ARG_UNUSED(code);
+}
+
+/* Called from xuk layer on actual CPU start */
+void _cpu_start(int cpu)
+{
+	xuk_set_f_ptr(cpu, &_kernel.cpus[cpu]);
+
+	/* Set up the timer ISR, but ensure the timer is disabled */
+	xuk_set_isr(INT_APIC_LVT_TIMER, 13, x86_apic_timer_isr, 0);
+	_apic.INIT_COUNT = 0;
+
+#ifdef CONFIG_IRQ_OFFLOAD
+	xuk_set_isr(XUK_INT_RAW_VECTOR(CONFIG_IRQ_OFFLOAD_VECTOR),
+		    -1, irq_offload_handler, 0);
+#endif
+
+	if (cpu <= 0) {
+		/* The SMP CPU startup function pointers act as init
+		 * flags.  Zero them here because this code is running
+		 * BEFORE .bss is zeroed!  Should probably move that
+		 * out of _Cstart() for this architecture...
+		 */
+		for (int i = 0; i < CONFIG_MP_NUM_CPUS; i++) {
+			cpu_init[i].fn = 0;
+		}
+
+		/* Enter Zephyr */
+		_Cstart();
+
+	} else if (cpu < CONFIG_MP_NUM_CPUS) {
+		/* SMP initialization.  First spin, waiting for
+		 * _arch_start_cpu() to be called from the main CPU
+		 */
+		while (!cpu_init[cpu].fn) {
+		}
+
+		/* Enter Zephyr, which will switch away and never return */
+		cpu_init[cpu].fn(0, cpu_init[cpu].arg);
+	}
+
+	/* Spin forever as a fallback */
+	while (1) {
+	}
+}
+
+/* Returns the initial stack to use for CPU startup on auxiliary (not
+ * cpu 0) processors to the xuk layer, which gets selected by the
+ * non-arch Zephyr kernel and stashed by _arch_start_cpu()
+ */
+unsigned int _init_cpu_stack(int cpu)
+{
+	return cpu_init[cpu].esp;
+}
+
+int _arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+			      void (*routine)(void *parameter), void *parameter,
+			      u32_t flags)
+{
+	ARG_UNUSED(flags);
+	__ASSERT(priority >= 2 && priority <= 15,
+		 "APIC interrupt priority must be 2-15");
+
+	xuk_set_isr(irq, priority, (void *)routine, parameter);
+	return 0;
+}
+
+void _arch_irq_disable(unsigned int irq)
+{
+	xuk_set_isr_mask(irq, 1);
+}
+
+void _arch_irq_enable(unsigned int irq)
+{
+	xuk_set_isr_mask(irq, 0);
+}
+
+void x86_apic_set_timeout(u32_t cyc_from_now)
+{
+	_apic.INIT_COUNT = cyc_from_now;
+}
+
+const NANO_ESF _default_esf;
+
+int x86_64_except_reason;
+
+void _NanoFatalErrorHandler(unsigned int reason, const NANO_ESF *esf)
+{
+	_SysFatalErrorHandler(reason, esf);
+}
+
+/* App-overridable handler.  Does nothing here */
+void __weak _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *esf)
+{
+	ARG_UNUSED(reason);
+	ARG_UNUSED(esf);
+	k_thread_abort(_current);
```

Comment:
I don't think it makes sense to do this if the crash occurs while in an ISR, since _current is just whatever thread was running when the IRQ happened.

Also, consider the case when k_panic() is called, that should really just hang.

Finally, I think you need a LOG_PANIC() call in here.

I don't care about K_ESSENTIAL that should just be removed from the kernel IMO.

Reply:
I don't really disagree, I'm just trying not to bake in what we "should" do when the truth is that almost everything is a cut-n-paste copy with different features dropped and added, and none of that gets tested anyway.   Will add an ISR check an a reason check for panic, sure.  Not sure what the expectations are for LOG stuff, will investigate.

And I think the failures are a cascade from the ARM loop.  A full sanitycheck runs for me offline, modulo the normal handful of timer-related failures spread across different architectures.

Reply:
> everything is a cut-n-paste copy with different features dropped and added, and none of that gets tested anyway

Definitely a fair point!

> And I think the failures are a cascade from the ARM loop

Yeah we need to address that, I'm not sure if this is some corner case in the ARM fault handling code or the kernel got corrupted in such an unlucky way that it induced that behavior. @ioannisg any thoughts on it? You can see here: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/29474/4/tests